dvc/ignore.py

```diff
@@ -23,25 +24,33 @@ def __call__(self, root, dirs, files):
 
 
 class DvcIgnorePatterns(DvcIgnore):
-    def __init__(self, ignore_file_path, tree):
-        assert os.path.isabs(ignore_file_path)
+    def __init__(self, pattern_list, dirname):
+
+        self.pattern_list = pattern_list
+        self.dirname = dirname
+        self.prefix = self.dirname + os.sep
 
-        self.ignore_file_path = ignore_file_path
-        self.dirname = os.path.normpath(os.path.dirname(ignore_file_path))
+        regex_pattern_list = map(
+            GitWildMatchPattern.pattern_to_regex, pattern_list
+        )
 
+        self.ignore_spec = [
+            (ignore, re.compile("|".join(item[0] for item in group)))
```

Comment:
Should one use `re.escape(item[0])` here?

Reply:
> It looks like you are reimplementing or completing implementing `GitWildMatchPattern`:
> 
> 1. Is this unavodable?
> 2. If it is then can this be made in `GitWildMatchPattern`?
> 3. If it can't then can it be separated from `dvc/ignore`, i.e. any logic not specific to dvc moved separately?

@Suor Thank you.
I learned a lot from your following advice. 
But, I didn't reimplement `GitWildMatchPattern`. What I'm doing is only changing the relative path of the `.dvcignore` file. And this is required if we merged two `.dvcignore` rules in different paths.
For example, if we had 
```
a
b
```
in  `/home/.dvcignore`
and 
```
!aa
!bb
```
in `/home/dir/.dvcignore`

And if we want to replace these two files in one.  We must translate one of them to another's directory.

We can choose:
1. change `/home/.dvcignore` into `/home/dir` but it is harder  
> But quicker I think because we can drop those rules which only influence other directories.  I was not as familiar with `gitignore` as now when I made this decision and overestimated the difficulty. I'd like to redo it this way this week.

2. change `/home/dir/.dvcignore` into `/home/` and then it would be
```
!/dir/**/aa
!/dir/**/bb
```

Reply:
`item[0]` are all generated by `GitWildMatchPattern` and it had been done in

https://github.com/cpburnz/python-path-specification/blob/3aee306db792bbabaf41d31f99441f94b1b3e740/pathspec/patterns/gitwildmatch.py#L278